pallets/providers/src/utils.rs

```diff
@@ -1097,38 +1112,67 @@ where
+                let delta_rate = new_bucket_rate
+                    .checked_sub(&bucket_rate)
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // If the rate has changed, update the payment stream.
+                if !delta_rate.is_zero() {
+                    // Since this is an increase, add the delta rate to the current rate.
+                    let new_rate = current_rate
+                        .checked_add(&delta_rate)
+                        .ok_or(ArithmeticError::Overflow)?;
 
-                <T::PaymentStreams as PaymentStreamsInterface>::update_fixed_rate_payment_stream(
+                    <T::PaymentStreams as PaymentStreamsInterface>::update_fixed_rate_payment_stream(
                     &msp_id, &user_id, new_rate,
-                )?;
+                	)?;
+                }
             }
             RateDeltaParam::Decrease(delta) => {
-                let delta_rate = value_prop
+                // Get the current bucket rate, which is the rate of a zero sized bucket plus the rate according to the bucket size.
+                let bucket_rate = value_prop
+                    .price_per_unit_of_data_per_block
+                    .checked_mul(&bucket.size.into())
+                    .ok_or(ArithmeticError::Overflow)?
+                    .checked_add(&T::ZeroSizeBucketFixedRate::get())
+                    .ok_or(ArithmeticError::Overflow)?;
+
+                // Calculate the new bucket's size.
+                let new_bucket_size = bucket
+                    .size
+                    .checked_sub(&delta)
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // Calculate what would be the new bucket rate with the new size.
+                let new_bucket_rate = value_prop
                     .price_per_unit_of_data_per_block
-                    .checked_mul(&delta.into())
+                    .checked_mul(&new_bucket_size.into())
+                    .ok_or(ArithmeticError::Overflow)?
+                    .checked_add(&T::ZeroSizeBucketFixedRate::get())
                     .ok_or(ArithmeticError::Overflow)?;
 
-                let new_rate = current_rate.saturating_sub(delta_rate);
+                // Get the delta rate, which is the difference between the old and new rates for this bucket.
+                let delta_rate = bucket_rate
+                    .checked_sub(&new_bucket_rate)
+                    .ok_or(ArithmeticError::Underflow)?;
+
+                // If the rate has changed, update the payment stream.
+                if !delta_rate.is_zero() {
+                    // Since this is a decrease, subtract the delta rate from the current rate.
+                    let new_rate = current_rate.saturating_sub(delta_rate);
 
-                if new_rate < T::ZeroSizeBucketFixedRate::get() {
-                    <T::PaymentStreams as PaymentStreamsInterface>::delete_fixed_rate_payment_stream(
```

Comment:
I think what had been suggested was to make the deposit be 0. Anyway, I wouldn't make anything free. If anything, I'd provide minimums.

Reply:
Payment streams should only be deleted when deleting the last bucket of it, otherwise there could be empty buckets lying around that are not charged even though we have a zero-sized bucket rate